web/src/queries/storage/config-model.ts

```diff
@@ -26,39 +26,129 @@ import { configModel } from "~/api/storage/types";
 import { QueryHookOptions } from "~/types/queries";
 import { SpacePolicyAction } from "~/types/storage";
 
+function copyModel(model: configModel.Config): configModel.Config {
+  return JSON.parse(JSON.stringify(model));
+}
+
+function isNewPartition(partition: configModel.Partition): boolean {
+  return partition.name === undefined;
+}
+
+function isSpacePartition(partition: configModel.Partition): boolean {
+  return partition.resizeIfNeeded || partition.delete || partition.deleteIfNeeded;
+}
+
+function isUsedPartition(partition: configModel.Partition): boolean {
+  return partition.filesystem !== undefined || partition.alias !== undefined;
+}
+
+function isReusedPartition(partition: configModel.Partition): boolean {
+  return !isNewPartition(partition) && isUsedPartition(partition) && !isSpacePartition(partition);
+}
+
 function findDrive(model: configModel.Config, driveName: string): configModel.Drive | undefined {
   const drives = model.drives || [];
   return drives.find((d) => d.name === driveName);
 }
 
-// TODO: add a second drive if needed (e.g., reusing a partition).
-function changeDrive(model: configModel.Config, driveName: string, newDriveName: string) {
+function removeDrive(model: configModel.Config, driveName: string) {
+  model.drives = model.drives.filter((d) => d.name !== driveName);
+}
+
+function isUsedDrive(model: configModel.Config, driveName: string) {
   const drive = findDrive(model, driveName);
-  if (drive === undefined) return;
+  if (drive === undefined) return false;
 
-  drive.name = newDriveName;
-  // TODO: assign space policy according to the use-case.
-  if (drive.spacePolicy === "custom") drive.spacePolicy = "keep";
+  return drive.partitions?.some((p) => isNewPartition(p) || isReusedPartition(p));
 }
 
-function setSpacePolicy(
-  model: configModel.Config,
+function isBoot(model: configModel.Config, driveName: string): boolean {
+  return model.boot?.configure && driveName === model.boot?.device?.name;
+}
+
+function isExplicitBoot(model: configModel.Config, driveName: string): boolean {
+  return !model.boot?.device?.default && driveName === model.boot?.device?.name;
+}
+
+function setBoot(originalModel: configModel.Config, boot: configModel.Boot) {
+  const model = copyModel(originalModel);
+  const name = model.boot?.device?.name;
+  const remove = name !== undefined && isExplicitBoot(model, name) && !isUsedDrive(model, name);
```

Comment:
Just to check I understood it correctly (I'm not against it). When a drive is not longer necessary after a change at the boot device, that drive is removed here. On the other hand, when a new drive needs to get added because a change at the boot device, that drive is added at the backend (during the conversion from the JSON model to config).

Right?